File: simple_deposit.py
#!/usr/bin/env python3
"""
Interacts with the app:
  python simple_deposit.py optin
  python simple_deposit.py deposit 1
  python simple_deposit.py info

Deposits must be grouped as [AppCall("deposit"), Payment], both from the same sender.
When your cumulative deposits reach 2 ALGO, the app inner-sends 2 ALGO back and logs "success".
"""

import json
import os
import sys
import base64
from algosdk.v2client import algod
from algosdk import mnemonic, account, transaction, logic
from algosdk.transaction import (
    ApplicationOptInTxn, ApplicationCallTxn, PaymentTxn, OnComplete, assign_group_id
)

# Configuration
ALGOD_URL = "https://testnet-api.algonode.cloud"
ALGOD_TOKEN = ""  # not needed for Algonode

# File paths
WALLET_PATH = os.path.join(os.path.dirname(__file__), "permanent_wallet.json")
DEPLOY_PATH = os.path.join(os.path.dirname(__file__), "simple_deployment.json")


def load_account():
    """Load account from permanent_wallet.json"""
    with open(WALLET_PATH, 'r') as f:
        wallet_data = json.load(f)
    
    # Get private key from mnemonic
    private_key = mnemonic.to_private_key(wallet_data['mnemonic'])
    derived_addr = account.address_from_private_key(private_key)
    
    if derived_addr != wallet_data['addr']:
        raise ValueError("Wallet address does not match mnemonic in permanent_wallet.json")
    
    return {"addr": derived_addr, "sk": private_key}


def load_deployment():
    """Load deployment info from simple_deployment.json"""
    with open(DEPLOY_PATH, 'r') as f:
        deploy_data = json.load(f)
    
    if 'appId' not in deploy_data:
        raise ValueError("simple_deployment.json missing appId")
    
    return deploy_data


def text_arg(s):
    """Convert string to bytes for app arguments"""
    return s.encode('utf-8')


def print_logs_from_tx(algod_client, tx_id):
    """Print logs and inner transactions from a transaction"""
    confirmed_txn = transaction.wait_for_confirmation(algod_client, tx_id, 4)
    
    # Print logs
    logs = confirmed_txn.get('logs', [])
    if logs:
        decoded_logs = [base64.b64decode(log).decode('utf-8') for log in logs]
        print("App Logs:", decoded_logs)
    
    # Print inner transactions (e.g., 2 ALGO refund)
    inner_txns = confirmed_txn.get('inner-txns', [])
    for i, itx in enumerate(inner_txns):
        if 'txn' in itx and itx['txn'].get('type') == 'pay':
            amount = itx['txn']['amt']
            receiver = itx['txn']['rcv']
            print(f"Inner payment #{i}: {amount} µALGO -> {receiver}")
    
    return confirmed_txn


def do_opt_in(algod_client, acct, app_id):
    """Opt into the application"""
    sp = algod_client.suggested_params()
    
    opt_in_txn = ApplicationOptInTxn(
        sender=acct['addr'],
        sp=sp,
        index=app_id
    )
    
    signed_txn = opt_in_txn.sign(acct['sk'])
    tx_id = algod_client.send_transaction(signed_txn)
    
    print(f"Sent opt-in tx: {tx_id}")
    transaction.wait_for_confirmation(algod_client, tx_id, 4)
    print("Opt-in confirmed")


def do_info(algod_client, acct, app_id):
    """Call the info function"""
    sp = algod_client.suggested_params()
    
    call_txn = ApplicationCallTxn(
        sender=acct['addr'],
        sp=sp,
        index=app_id,
        on_complete=OnComplete.NoOpOC,
        app_args=[text_arg("info")]
    )
    
    signed_txn = call_txn.sign(acct['sk'])
    tx_id = algod_client.send_transaction(signed_txn)
    
    print(f"Sent info call tx: {tx_id}")
    print_logs_from_tx(algod_client, tx_id)


def do_deposit(algod_client, acct, app_id, algos):
    """Make a deposit to the application"""
    if not isinstance(algos, (int, float)) or algos <= 0:
        raise ValueError("Deposit amount must be a positive number of ALGOs")
    
    amount = round(algos * 1_000_000)  # Convert to microALGO
    
    sp = algod_client.suggested_params()
    app_addr = logic.get_application_address(app_id)
    
    # 1) App call first
    app_call_txn = ApplicationCallTxn(
        sender=acct['addr'],
        sp=sp,
        index=app_id,
        on_complete=OnComplete.NoOpOC,
        app_args=[text_arg("deposit")]
    )
    
    # 2) Payment second
    pay_txn = PaymentTxn(
        sender=acct['addr'],
        sp=sp,
        receiver=app_addr,
        amt=amount
    )
    
    # Group transactions
    assign_group_id([app_call_txn, pay_txn])
    
    # Sign transactions
    signed_app_call = app_call_txn.sign(acct['sk'])
    signed_payment = pay_txn.sign(acct['sk'])
    
    # Send grouped transaction
    tx_id = algod_client.send_transactions([signed_app_call, signed_payment])
    print(f"Sent grouped deposit txs. Group ID (first tx): {tx_id}")
    
    print_logs_from_tx(algod_client, tx_id)


def main():
    """Main function"""
    if len(sys.argv) < 2:
        print("Usage:")
        print("  python simple_deposit.py optin")
        print("  python simple_deposit.py deposit 1")
        print("  python simple_deposit.py info")
        sys.exit(1)
    
    cmd = sys.argv[1]
    if cmd not in ["optin", "deposit", "info"]:
        print("Invalid command. Use: optin, deposit, or info")
        sys.exit(1)
    
    try:
        # Initialize client and load data
        algod_client = algod.AlgodClient(ALGOD_TOKEN, ALGOD_URL)
        acct = load_account()
        deployment = load_deployment()
        app_id = deployment['appId']
        
        # Execute command
        if cmd == "optin":
            do_opt_in(algod_client, acct, app_id)
        elif cmd == "info":
            do_info(algod_client, acct, app_id)
        elif cmd == "deposit":
            if len(sys.argv) < 3:
                raise ValueError("Please provide an amount in ALGO, e.g. 'deposit 1'")
            amount = float(sys.argv[2])
            do_deposit(algod_client, acct, app_id, amount)
    
    except Exception as e:
        error_msg = str(e).lower()
        if "already opted in" in error_msg:
            print("Already opted-in. You can proceed to deposit.")
            return
        
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: simple_deposit.js
// Interacts with the app:
//   node simple_deposit.js optin
//   node simple_deposit.js deposit 1
//   node simple_deposit.js info
//
// Deposits must be grouped as [AppCall("deposit"), Payment], both from the same sender.
// When your cumulative deposits reach 2 ALGO, the app inner-sends 2 ALGO back and logs "success".

const fs = require("fs");
const path = require("path");
const algosdk = require("algosdk");

const ALGOD_URL = "https://testnet-api.algonode.cloud";
const ALGOD_TOKEN = ""; // not needed for Algonode
const ALGOD_PORT = "";

const WALLET_PATH = path.join(__dirname, "permanent_wallet.json");
const DEPLOY_PATH = path.join(__dirname, "simple_deployment.json");

// Normalize address across SDK v3/v4
function deriveAddr(acct) {
  if (typeof acct.addr === "string") return acct.addr;
  if (acct.addr && acct.addr.publicKey instanceof Uint8Array) {
    return algosdk.encodeAddress(acct.addr.publicKey);
  }
  return String(acct.addr);
}

function loadAccount() {
  const j = JSON.parse(fs.readFileSync(WALLET_PATH, "utf8"));
  const acct = algosdk.mnemonicToSecretKey(j.mnemonic);

  const derivedAddr = deriveAddr(acct);
  if (derivedAddr !== j.addr) {
    throw new Error("Wallet address does not match mnemonic in permanent_wallet.json");
  }
  return { addr: derivedAddr, sk: acct.sk };
}

function loadDeployment() {
  const j = JSON.parse(fs.readFileSync(DEPLOY_PATH, "utf8"));
  if (!j.appId) throw new Error("simple_deployment.json missing appId");
  return j;
}

function textArg(s) {
  return new Uint8Array(Buffer.from(s));
}

async function printLogsFromTx(algod, txId) {
  const conf = await algosdk.waitForConfirmation(algod, txId, 4);
  const logs = conf.logs || [];
  const dec = logs.map((b) => Buffer.from(b, "base64").toString("utf8"));
  if (dec.length) console.log("App Logs:", dec);
  // Show inner payments (e.g. 2 ALGO refund)
  if (conf["inner-txns"]) {
    conf["inner-txns"].forEach((itx, i) => {
      if (itx.txn && itx.txn.txn && itx.txn.txn.type === "pay") {
        const amt = itx.txn.txn.amt;
        const rcvr = algosdk.encodeAddress(new Uint8Array(itx.txn.txn.rcv));
        console.log(`Inner payment #${i}: ${amt} µALGO -> ${rcvr}`);
      }
    });
  }
  return conf;
}

async function doOptIn(algod, acct, appId) {
  const sp = await algod.getTransactionParams().do();
  const optin = algosdk.makeApplicationOptInTxnFromObject({
    from: acct.addr,
    appIndex: appId,
    suggestedParams: sp,
  });
  const signed = optin.signTxn(acct.sk);
  const { txId } = await algod.sendRawTransaction(signed).do();
  console.log("Sent opt-in tx:", txId);
  await algosdk.waitForConfirmation(algod, txId, 4);
  console.log("Opt-in confirmed");
}

async function doInfo(algod, acct, appId) {
  const sp = await algod.getTransactionParams().do();
  const call = algosdk.makeApplicationCallTxnFromObject({
    from: acct.addr,
    appIndex: appId,
    onComplete: algosdk.OnApplicationComplete.NoOpOC,
    suggestedParams: sp,
    appArgs: [textArg("info")],
  });
  const signed = call.signTxn(acct.sk);
  const { txId } = await algod.sendRawTransaction(signed).do();
  console.log("Sent info call tx:", txId);
  await printLogsFromTx(algod, txId);
}

async function doDeposit(algod, acct, appId, algos) {
  if (!Number.isFinite(algos) || algos <= 0) {
    throw new Error("Deposit amount must be a positive number of ALGOs");
  }
  const amount = Math.round(algos * 1_000_000);

  const sp = await algod.getTransactionParams().do();
  const appAddr = algosdk.getApplicationAddress(appId);

  // 1) App call first
  const appCallTxn = algosdk.makeApplicationCallTxnFromObject({
    from: acct.addr,
    appIndex: appId,
    onComplete: algosdk.OnApplicationComplete.NoOpOC,
    suggestedParams: sp,
    appArgs: [textArg("deposit")],
  });

  // 2) Payment second
  const payTxn = algosdk.makePaymentTxnWithSuggestedParamsFromObject({
    from: acct.addr,
    to: appAddr,
    amount,
    suggestedParams: sp,
  });

  algosdk.assignGroupID([appCallTxn, payTxn]);

  const signed1 = appCallTxn.signTxn(acct.sk);
  const signed2 = payTxn.signTxn(acct.sk);

  const { txId } = await algod.sendRawTransaction([signed1, signed2]).do();
  console.log(`Sent grouped deposit txs. Group ID (first tx): ${txId}`);

  await printLogsFromTx(algod, txId);
}

async function main() {
  const [cmd, arg] = process.argv.slice(2);
  if (!cmd || !["optin", "deposit", "info"].includes(cmd)) {
    console.log("Usage:");
    console.log("  node simple_deposit.js optin");
    console.log("  node simple_deposit.js deposit 1");
    console.log("  node simple_deposit.js info");
    process.exit(1);
  }

  const algod = new algosdk.Algodv2(ALGOD_TOKEN, ALGOD_URL, ALGOD_PORT);
  const acct = loadAccount();
  const { appId } = loadDeployment();

  try {
    if (cmd === "optin") {
      await doOptIn(algod, acct, appId);
    } else if (cmd === "info") {
      await doInfo(algod, acct, appId);
    } else if (cmd === "deposit") {
      const amt = parseFloat(arg);
      if (!arg) throw new Error("Please provide an amount in ALGO, e.g. `deposit 1`");
      await doDeposit(algod, acct, appId, amt);
    }
  } catch (e) {
    const msg = String(e && e.message ? e.message : e);
    if (msg.toLowerCase().includes("already opted in")) {
      console.log("Already opted-in. You can proceed to deposit.");
      return;
    }
    console.error(e);
    process.exit(1);
  }
}

main();